README.md
[![npm version](https://badge.fury.io/js/whoots-js.svg)](https://badge.fury.io/js/whoots-js)
[![Build Status](https://circleci.com/gh/mapbox/whoots-js.svg?style=svg)](https://circleci.com/gh/mapbox/whoots-js)

## whoots-js

Request tiles from WMS servers that support EPSG:3857.

This project is a JavaScript port of https://github.com/timwaters/whoots by Tim Waters.


### What is it?

Given a `z/x/y` tile coordinate like `19/154308/197167`, `whoots-js` can request imagery from an EPSG:3857 supporting WMS server like this:

```
http://geodata.state.nj.us/imagerywms/Natural2015?
  bbox=-8242663.382160267,4966572.349857613,-8242586.945131982,4966648.786885899
  &format=image/png&service=WMS&version=1.1.1&request=GetMap&srs=EPSG:3857
  &width=256&height=256&layers=Natural2015
```


### Usage

```js
var wms = new WhooTS();

// Get an image url for a given tile coordinate
var baseUrl = 'http://geodata.state.nj.us/imagerywms/Natural2015';
var layer = 'Natural2015';
var url = whoots.getUrl(baseUrl, layer, 154308, 197167, 19);
```


### Server

This project includes a sample redirecting wms proxy server in `server.js`.

`npm run server` will start a local server on port 8080 that redirects tile requests.

Valid tile requests look like:

```
http://localhost:8080/tms/{z}/{x}/{y}/{layer}/{endpoint}
http://localhost:8080/tms/19/154308/197167/Natural2015/http://geodata.state.nj.us/imagerywms/Natural2015
```


### Documentation

Complete API documentation is here:  http://mapbox.github.io/whoots-js/docs/


CHANGELOG.md
:warning: = Breaking change

## 2.0.0
##### Jun 1, 2016
* :warning: Refactor for a classless API

## 1.1.0
##### May 23, 2016
* Make getTileBbox and getMercCoords public

## 1.0.0
##### May 23, 2016
* Initial release


test/index.js
'use strict';

var test = require('tap').test;
var WhooTS = require('../.');


test('WhooTS', function(t) {
    t.test('constructor', function(t) {
        var wms = new WhooTS();
        t.type(wms, 'object');
        t.end();
    });

    t.test('getUrl', function(t) {
        var wms = new WhooTS();
        var baseUrl = 'http://geodata.state.nj.us/imagerywms/Natural2015';
        var layer = 'Natural2015';
        var url = wms.getUrl(baseUrl, layer, 154308, 197167, 19);
        t.deepEqual(url, 'http://geodata.state.nj.us/imagerywms/Natural2015?' +
            'bbox=-8242663.382160267,4966572.349857613,-8242586.945131982,4966648.786885899' +
            '&format=image/png' +
            '&service=WMS' +
            '&version=1.1.1' +
            '&request=GetMap' +
            '&srs=EPSG:3857' +
            '&width=256' +
            '&height=256' +
            '&layers=Natural2015'
        );
        t.end();
    });

    t.test('getTileBbox', function(t) {
        var wms = new WhooTS();
        var bbox = wms.getTileBbox(154308, 197167, 19);
        t.deepEqual(bbox, '-8242663.382160267,4966572.349857613,-8242586.945131982,4966648.786885899');
        t.end();
    });

    t.test('getMercCoords', function(t) {
        var wms = new WhooTS();
        var coords = wms.getMercCoords(154308 * 256, 327120 * 256, 19);
        t.deepEqual(coords, [-8242663.382160267, 4966572.349857613]);
        t.end();
    });

    t.end();
});


server.js
'use strict';

var WhooTS = require('.');
var http = require('http');
var url = require('url');

var PORT = 8080;

// valid requests look like:
// http://localhost:8080/tms/{z}/{x}/{y}/{layer}/{baseUrl}
// http://localhost:8080/tms/19/154308/197167/Natural2015/http://geodata.state.nj.us/imagerywms/Natural2015
function handleRequest(request, response) {
    var pathname = url.parse(request.url, true).pathname;
    var params = pathname.split('/');

    if (params.length > 6 && params[1].toLowerCase() === 'tms') {
        var z = +params[2],
            x = +params[3],
            y = +params[4],
            layer = params[5],
            baseUrl = pathname.replace(params.slice(0,6).join('/') + '/', '');

        if (!isNaN(z) && !isNaN(x) && !isNaN(y) && layer && url.parse(baseUrl).protocol) {
            var wms = new WhooTS();
            response.writeHead(302, { 'Location': wms.getUrl(baseUrl, layer, x, y, z) });
            response.end('Redirect');
            return;
        }
    }

    response.statusCode = '404';
    response.end('Not Found');
}


var server = http.createServer(handleRequest);
server.listen(PORT, function() {
    /* eslint-disable no-console */
    console.log('Server listening on: http://localhost:%s', PORT);
    /* eslint-enable no-console */
});


index.js
'use strict';

/**
 * getURL
 *
 * @param    {String}  baseUrl  Base url of the WMS server
 * @param    {String}  layer    Layer name
 * @param    {Number}  x        Tile coordinate x
 * @param    {Number}  y        Tile coordinate y
 * @param    {Number}  z        Tile zoom
 * @param    {Object}  [options]
 * @param    {String}  [options.format='image/png']
 * @param    {String}  [options.service='WMS']
 * @param    {String}  [options.version='1.1.1']
 * @param    {String}  [options.request='GetMap']
 * @param    {String}  [options.srs='EPSG:3857']
 * @param    {Number}  [options.width='256']
 * @param    {Number}  [options.height='256']
 * @returns  {String}  url
 * @example
 * var baseUrl = 'http://geodata.state.nj.us/imagerywms/Natural2015';
 * var layer = 'Natural2015';
 * var url = whoots.getURL(baseUrl, layer, 154308, 197167, 19);
 */
exports.getURL = function(baseUrl, layer, x, y, z, options) {
    options = options || {};

    var url = baseUrl + '?' + [
        'bbox='    + exports.getTileBBox(x, y, z),
        'format='  + (options.format || 'image/png'),
        'service=' + (options.service || 'WMS'),
        'version=' + (options.version || '1.1.1'),
        'request=' + (options.request || 'GetMap'),
        'srs='     + (options.srs || 'EPSG:3857'),
        'width='   + (options.width || 256),
        'height='  + (options.height || 256),
        'layers='  + layer
    ].join('&');

    return url;
};


/**
 * getTileBBox
 *
 * @param    {Number}  x  Tile coordinate x
 * @param    {Number}  y  Tile coordinate y
 * @param    {Number}  z  Tile zoom
 * @returns  {String}  String of the bounding box
 */
exports.getTileBBox = function(x, y, z) {
    // for Google/OSM tile scheme we need to alter the y
    y = (Math.pow(2, z) - y - 1);

    var min = exports.getMercCoords(x * 256, y * 256, z),
        max = exports.getMercCoords((x + 1) * 256, (y + 1) * 256, z);

    return min[0] + ',' + min[1] + ',' + max[0] + ',' + max[1];
};

/**
 * getMercCoords
 *
 * @param    {Number}  x  Pixel coordinate x
 * @param    {Number}  y  Pixel coordinate y
 * @param    {Number}  z  Tile zoom
 * @returns  {Array}   [x, y]
 */
exports.getMercCoords = function(x, y, z) {
    var resolution = (2 * Math.PI * 6378137 / 256) / Math.pow(2, z),
        merc_x = (x * resolution - 2 * Math.PI  * 6378137 / 2.0),
        merc_y = (y * resolution - 2 * Math.PI  * 6378137 / 2.0);

    return [merc_x, merc_y];
};
